proposed/dynamic-component-loading.md

```diff
@@ -0,0 +1,110 @@
+    * This provides full isolation to the component. Every dependency the component carries with it will be used from the component and loaded in isolation from the rest of the app. Basically avoids any potential collisions.
+    * On the downside, this doesn't provide implicit sharing. Typically only framework assemblies would be shared in this scenario. Component would have to explicitly choose which assemblies to share, by not carrying them with it (this can be setup in project file by using `CopyLocal=false` for assembly references, similar option exists for project and NuGet references as well). This means that types used for communication between the app and the component would have to be explicitly shared by the component (via the exclusion of the assembly in the component). This is not done by the SDK by default, so it's easy to get this wrong and even with improved diagnostics will be relatively hard to debug.
+* **Always load into default** - in this case all loads are done to the parent (Default) load context. In the extreme case a new load context is not really needed. Alternative could be that the load is attempted to the default load context (first by name, then by resolved file path). If that fails (should really only happen in case of a collision), the dependency would be loaded into the new load context in isolation.
+    * Inherently shares as much as possible - avoids problems of sharing types used for communication.
+    * Downside is that loading a component will "pollute" the default load context with assemblies from the component.
+    * Loading multiple components with similar dependencies can easily lead to unpredictable results as what gets loaded where would depend on ordering.
+    * Auto-upgrades - if the app uses a higher version of a given dependency, component which uses a lower version of the dependency will get the app's version - this can lead to incompatibilities.
+* **Prefer default** - in this case a new load context is created. When it tries to resolve a dependency, it will first try to resolve it against the parent (Default) load context. If the parent can satisfy the dependency, it will be used. Otherwise if the dependency can't be resolved by the parent, the dependency will be resolved to a file path and loaded in isolation into the new load context.
+    * Inherently shares all assemblies from the parent (Default) - avoids problems with sharing types used for communication.
+    * There's no pollution of the parent context - no new files will be loaded into the parent context.
+    * Auto-upgrade - components will auto-upgrade to the version of a dependency they share with the app (downgrade will not occur, in that case the dependency would be loaded in isolation). This can lead to incompatibilities.
+
+Open questions:
```

Comment:
Important consideration for me is versioning of EventSource names and manifests when multiple versions are loaded into the process. Same for diagnostics source.

It is my understanding that all isolation models will allow to subscribe to event and diagnostics sources from the loaded assembly as those sources are defined in system assemblies. If so - some naming issue may occur.

CC: @vancem

Reply:
Please note that isolation is not a new feature. It has been available for a long time now. It's done through the `AssemblyLoadContext` class which is not new. Also some methods on `Assembly` provide isolation (`Assembly.LoadFile` loads the specified file in isolation). So it's already possible to load even the same assembly twice (or two similar assemblies and so on). I don't know if there's been any thinking on the impact of this capability on the EventSource mechanisms. I'll let @vancem or @brianrob to comment on that.

Reply:
@vancem, @brianrob what do you think? Can you confirm that loading in separate `AssemblyLoadContext` will still make everybody use the "global" Event and Diagnostics Source?

Reply:
@SergeyKanzhelev  - I view all this loader stuff as pretty orthogonal to event/source naming issues in EventSource/DiagnosticSource.  (Yes there are issues but this feature does not make the problem any better or worse).

The one place I can think of that is a issue is that TODAY we make it illegal to create two EventSources with the same name in the same AppDomain.   We did this for relatively weak reasons (it is a perf issue).   This behavior WILL cause grief if an assembly is loaded twice (because of isolation).   It will call the SHARED System.Private.Corelib version of EventSource which will fail because the name is already used.    I have always been ambivalent about this rule, so I don't mind if we stop checking for it. 

As far as the rest goes, I think it is roughly orthogonal.   The names WILL overlap but that is good (you think of them as the 'same' events really).   For DiagnosticSource, there will be a separate publication point for each System.Diagnostic.DiagnosticSource assembly that is loaded (since there is a static in their that you subscribe to to get events.  

So we may wish to relax the uniqueness check in EventSource, but otherwise I think it will be OK.   If you have specific concerns, we can walk through specific cases in more detail.

Reply:
>  This behavior WILL cause grief if an assembly is loaded twice (because of isolation). It will call the SHARED System.Private.Corelib version of EventSource which will fail because the name is already used. 

So custom EventSource creation will crash? Potentially an entire process? Or this will be handled? In any case - this may become a common case with plugins using libraries like Application Insights for telemetry reporting. Application Insights SDK instantiate an `EventSource` for troubleshooting.

> For DiagnosticSource, there will be a separate publication point for each System.Diagnostic.DiagnosticSource assembly that is loaded

Will different `DiagnosticSource`-s share the AsyncLocal (`Activity.Current`)? If not - distributed tracing scenarios like Activity started in ASP.NET core host and than child activity created inside the plugin would not work. Is there anything that can be done to special-case the DiagnosticsSource?